### PR TITLE
Test interface

### PR DESCRIPTION
**Pull request by Duhemm**
This is a super pre alpha version of the test interface. I'm opening the PR for people who may want to play with it, and look at what it may look like in the future. It still has quite a few issues. It's certainly not worth doing a full review at the moment.

If you want to try it out, you'll need to:

 * checkout that branch and `publishLocal`ly it first: https://github.com/Duhemm/utest/tree/cross-native
 * and then `rebuild` this repository (this should take care of publishing everything in the right order).

It follows the design of the test interface of ScalaJS, but there are a few differences:

 * The distinction between master and slave is still half there. The distinction is not observed at the moment, as it is not clear to me why this is needed.
 * We don't serialise the task, but keep them on the remote runner and refer to them using an ID that is assigned.
 * We have a different serialisation.

### CODE REVIEW

**Comment by muxanick** [comment]
@Duhemm great work, do you mind to move sockets to separate PR? I would like to add windows port

**Comment by Duhemm** [comment]
@muxanick I just opened #758

**Comment by Duhemm** [comment]
I just tested, it works nice with uTest and scalacheck (modulo fixing a compiler bug for scalacheck):

```
> sandbox/test
[info] Linking (1937 ms)
[info] Discovered 1847 classes and 12207 methods
[info] Optimizing (1503 ms)
[info] Generating intermediate code (403 ms)
[info] Produced 61 files
[info] Compiling to native code (2271 ms)
[info] Linking native code (312 ms)
[info] Starting process '/Users/martin/Documents/Projects.nosync/Duhemm/scala-native/sandbox/target/scala-2.11/sandbox-out' on port '9841'.
[info] Client connected with IP address: 0.0.0.0
[info] Starting process '/Users/martin/Documents/Projects.nosync/Duhemm/scala-native/sandbox/target/scala-2.11/sandbox-out' on port '9363'.
[info] Client connected with IP address: 0.0.0.0
[info] -------------------------Starting Suite foobar.MoreTest-------------------------
[info] foobar.MoreTest.test2            Success
[info] foobar.MoreTest          Success
[info] ------------------------Starting Suite foobar.UTestTest------------------------
[info] Printing something from test...
[info] foobar.UTestTest.test1           Success
[info] foobar.UTestTest         Success
[info] + String.startsWith: OK, passed 100 tests.
[info] ScalaCheck
[info] Passed: Total 1, Failed 0, Errors 0, Passed 1
[info] utest
[info] -----------------------------------Results-----------------------------------
[info] foobar.UTestTest         Success
[info]     test1                Success
[info] foobar.MoreTest          Success
[info]     test2                Success
[info]
[info] Tests: 4
[info] Passed: 4
[info] Failed: 0
[info] Passed: Total 0, Failed 0, Errors 0, Passed 0
```

**Comment by densh** [comment]
Needs a rebase due to conflicting changes in .travis.yml

File: .travis.yml [hunk 1]
```diff
@@ -67,7 +67,24 @@ script:
   - java -version
   - if [ -n "$CLANG_FORMAT_PATH" ]; then bin/clangfmt --test; fi
   - bin/scalafmt --test
-  - bash sbt -J-Xmx2G rebuild "set scriptedBufferLog in sbtScalaNative := false" test-all
+  - sbt -J-Xmx2G rebuild
+  - mkdir utest
+  - pushd utest
+  - git init .
+  - git remote add -t cross-native Duhemm https://github.com/Duhemm/utest.git
```
Comment (densh) [inline .travis.yml]: For the testing frameworks: they should depend on the test interface artifact, not the stubs artifact.

Comment (densh) [inline .travis.yml]: Also, once this PR is merged we should upstream changes made to `utest` and `scalacheck`.

File: build.sbt [hunk 2]
```diff
@@ -483,3 +509,39 @@ lazy val testingCompiler =
...
+    .settings(toolSettings)
+    .settings(scalaVersion := libScalaVersion)
+    .in(file("test-interface"))
+    .settings(
+      name := "test-interface",
+      libraryDependencies += "org.scala-sbt"    % "test-interface"   % "1.0",
+      libraryDependencies -= "org.scala-native" %%% "test-interface" % version.value % Test
+    )
+    .enablePlugins(ScalaNativePlugin)
+    .dependsOn(testInterfaceSerialization, stubs)
```
Comment (densh) [inline build.sbt]: Do we really need to split test interface and stubs sub projects? Can we merge them together?

File: .travis.yml (hunk 1 above)
Comment (densh) [inline .travis.yml]: Btw, sbt-crossproject is now at `0.2.0`, we released newer version but haven't announced it yet.

**Comment by densh** [comment]
@Duhemm Sockets are merged now, please rebase on top of latest master to resolve the merge conflict.

File: scalanative-stubs/src/main/scala/scala/scalanative/testinterface/PreLoadedClassLoader.scala [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package scala.scalanative.testinterface
+
+/** A `ClassLoader` that returns pre-instanciated objects.
+ *  We use it to bypass uses of reflection in the test interface.
+ */
+class PreLoadedClassLoader(map: Map[String, AnyRef]) extends ClassLoader {
```
Comment (densh) [inline scalanative-stubs/src/main/scala/scala/scalanative/testinterface/PreLoadedClassLoader.scala]: Given that `preload` is a single word, the name should be `PreloadedClassLoader`.

**Comment by Duhemm** [comment]
@densh I've rebased it. It still includes everything to clone and build my forks of utest and scalacheck. Let me know when you think this PR is ready so that I can remove all of that.

File: test-interface-serialization/src/main/scala/scala/scalanative/testinterface/serialization/Serializer.scala [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import scala.compat.Platform.EOL
+import sbt.testing._
+
+import scala.collection.mutable.ListBuffer
+
+object Serializer {
+  def serialize[T: Serializable](v: T): Iterator[String] =
+    implicitly[Serializable[T]].serialize(v)
+  def deserialize[T: Serializable](in: Iterator[String]): T =
+    implicitly[Serializable[T]].deserialize(in)
```
Comment (densh) [inline test-interface-serialization/src/main/scala/scala/scalanative/testinterface/serialization/Serializer.scala]: Given that serialization here isn't meant to be extensible, it's better to go for lower-level abstraction-free approach here. Current implementation is probably rather slow as it relies on virtual dispatch heavily.

First of all type-classes are an overkill here due to the fact that all instances are statically known and we don't expect any extensibility beyond that point.

Second, for serialization it's best to have a mutable builder-like code to minimize I/O overhead. Please have a look at [NIR serialization](https://github.com/scala-native/scala-native/tree/master/nir/src/main/scala/scala/scalanative/nir/serialization) and [Show builder](https://github.com/scala-native/scala-native/blob/master/util/src/main/scala/scala/scalanative/util/Show.scala) for examples. We used to do high-level type-class like code and it was very expensive, even on the JVM (2-3x performance overhead.)

File: nativelib/src/main/scala/java/lang/reflect/Field.scala [hunk 5]
```diff
@@ -0,0 +1,5 @@
+package java.lang.reflect
+
+class Field {
+  def get(obj: Object): Object = null
```
Comment (densh) [inline nativelib/src/main/scala/java/lang/reflect/Field.scala]: The same problem as for `Class::getField`, it should throw via `???` given that it's not implemented.

File: scalanative-stubs/src/main/scala/scala/scalajs/js/annotation/JSExportDescendentObjects.scala [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package scala.scalajs.js.annotation
+
+class JSExportDescendentObjects(ignoreInvalidDescendants: Boolean)
```
Comment (densh) [inline scalanative-stubs/src/main/scala/scala/scalajs/js/annotation/JSExportDescendentObjects.scala]: This is one is deprecated in Scala.js and we should keep it being deprecated in native.

File: scalanative-stubs/src/main/scala/scala/scalajs/js/annotation/JSExportDescendentClasses.scala [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package scala.scalajs.js.annotation
+
+class JSExportDescendentClasses(ignoreInvalidDescendants: Boolean)
```
Comment (densh) [inline scalanative-stubs/src/main/scala/scala/scalajs/js/annotation/JSExportDescendentClasses.scala]: This is one is deprecated in Scala.js and we should keep it being deprecated in native.

File: test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+      scala.sys.exit(err)
+    }
+  }
+
+  /** Sets up the server socket, returns the file descriptor. */
+  private def setupServer(port: uint16_t)(implicit zone: Zone): CInt = {
+    val server_address = native.alloc[sockaddr_in]
+    !server_address._1 = socket.AF_INET.toUShort
+    !server_address._2 = htons(port)
+    !server_address._3._1 = htonl(INADDR_ANY)
```
Comment (densh) [inline test-interface/src/main/scala/scala/scalanative/testinterface/TestMainBase.scala]: sockaddr and friends badly need to have implicit classes that alias their field names to their human-readable counter-parts.

**Comment by Duhemm** [comment]
@densh I've address all of your comments.

File: .travis.yml [hunk 9]
```diff
@@ -61,7 +61,24 @@ script:
...
+  - popd
+  - mkdir scalacheck
+  - pushd scalacheck
+  - git init .
+  - git remote add -t cross-native Duhemm https://github.com/Duhemm/scalacheck.git
+  - git fetch Duhemm cross-native
+  - git checkout Duhemm/cross-native
+  - sbt publishLocal
+  - popd
+  - bash sbt -J-Xmx2G "set scriptedBufferLog in sbtScalaNative := false" sandbox/test test-all
```
Comment (densh) [inline .travis.yml]: As soon as we remove temporary dependency on your branch of `utest` and `scalacheck`, we'll have no tests whatsoever for the testing framework integration. I suggest that we migrate our current mini-framework we used for `tests` project to the testing framework integration to address this.

File: unit-tests/src/main/scala/tests/Main.scala [hunk 10]
```diff
@@ -1,24 +1,24 @@
...
+//      success = success && suiteSuccess
+//    }
+//
+//    if (!success) {
+//      println("Some tests failed. See above for details.")
+//    }
+//
+//    exit(if (success) 0 else 1)
+//  }
+//}
```
Comment (densh) [inline unit-tests/src/main/scala/tests/Main.scala]: This file is not necessary any longer.

**Comment by Duhemm** [comment]
@densh Done!

**Comment by densh** [comment]
Another conflicting change due to PR merged to master on `FileInputStreamSuite.scala`...

**Comment by Duhemm** [comment]
@densh Done!

**Comment by densh** [comment]
As soon as CI passes, this is LGTM! Fantastic work @Duhemm! ✨ 🍰 ✨

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/TestUtilities.scala [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+  private def makeTestsMap(tests: Seq[TestDefinition]): String =
+    tests
+      .map { t =>
+        val isModule = t.fingerprint match {
+          case af: AnnotatedFingerprint => af.isModule
+          case sf: SubclassFingerprint  => sf.isModule
+        }
+
+        val inst = if (isModule) t.name else s"new ${t.name}"
```
Comment (xuwei-k) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/TestUtilities.scala]:
```scala
val inst = if (isModule) s"_root_.${t.name}" else s"new _root_.${t.name}"
s""""${t.name}" -> $inst"""
```

https://github.com/rickynils/scalacheck/pull/337#issuecomment-308925196

**Pull request merged**